press/press/doctype/deploy_candidate_build/deploy_candidate_build.py

```diff
@@ -0,0 +1,1167 @@
+}
+
+
+def get_build_stage_and_step(
+	stage_slug: str, step_slug: str, app_titles: dict[str, str] | None = None
+) -> tuple[str, str]:
+	stage = STAGE_SLUG_MAP.get(stage_slug, stage_slug)
+	step = step_slug
+	if stage_slug == "clone" or stage_slug == "apps":
+		step = app_titles.get(step_slug, step_slug)
+	else:
+		step = STEP_SLUG_MAP.get((stage_slug, step_slug), step_slug)
+	return (stage, step)
+
+
+def get_remote_step_output(
+	step_name: typing.Literal["build", "push"],
+	output_data: dict,
+	response_data: dict | None,
+):
+	if output := output_data.get(step_name):
+		return output
+
+	if not isinstance(response_data, dict):
+		return None
+
+	job_step_name = "Build Image" if step_name == "build" else "Push Docker Image"
+	for step in response_data.get("steps", []):
+		if step.get("name") != job_step_name:
+			continue
+
+		commands = step.get("commands", [])
+		if not isinstance(commands, list) or len(commands) == 0 or not isinstance(commands[0], dict):
+			continue
+
+		output = commands[0].get("output")
+		if not isinstance(output, str):
+			continue
+
+		with contextlib.suppress(AttributeError, json.JSONDecodeError):
+			return json.loads(output).get(step_name, [])
+
+	return None
+
+
+def get_duration(start_time: datetime, end_time: datetime | None = None):
+	end_time = end_time or now()
+	seconds_elapsed = (end_time - start_time).total_seconds()
+	value = rounded(seconds_elapsed, 3)
+	return float(value)
+
+
+class DeployCandidateBuild(Document):
+	# begin: auto-generated types
+	# This code is auto-generated. Do not modify anything in this block.
+
+	from typing import TYPE_CHECKING
+
+	if TYPE_CHECKING:
+		from frappe.types import DF
+
+		from press.press.doctype.deploy_candidate_build_step.deploy_candidate_build_step import (
+			DeployCandidateBuildStep,
+		)
+
+		build_directory: DF.Data | None
+		build_duration: DF.Time | None
+		build_end: DF.Datetime | None
+		build_error: DF.Code | None
+		build_output: DF.Code | None
+		build_server: DF.Link | None
+		build_start: DF.Datetime | None
+		build_steps: DF.Table[DeployCandidateBuildStep]
+		deploy_after_build: DF.Check
+		deploy_candidate: DF.Link
+		docker_image: DF.Data | None
+		docker_image_id: DF.Data | None
+		docker_image_repository: DF.Data | None
+		docker_image_tag: DF.Data | None
+		error_key: DF.Data | None
+		group: DF.Link
+		manually_failed: DF.Check
+		no_build: DF.Check
+		no_cache: DF.Check
+		no_push: DF.Check
+		pending_duration: DF.Time | None
+		pending_end: DF.Datetime | None
+		pending_start: DF.Datetime | None
+		retry_count: DF.Int
+		scheduled_time: DF.Datetime | None
+		status: DF.Literal["Draft", "Scheduled", "Pending", "Preparing", "Running", "Success", "Failure"]
+		user_addressable_failure: DF.Check
+	# end: auto-generated types
+
+	dashboard_fields = (
+		"name",
+		"status",
+		"creation",
+		"deployed",
+		"build_steps",
+		"build_start",
+		"build_end",
+		"build_duration",
+		"build_error",
+		"group",
+		"retry_count",
+	)
```

Comment:
For `dashboard_fields` to work, you should have the `team` field in doctype.
That acts for authentication and filter.

Reply:
Added